### ISSUE TITLE
F* allows declaring the same top-level identifier more than once

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#1106]
F* (both batch and interactive mode) allows to declare the same top-level identifier more than once. This is very confusing, since only the last declaration seems to be used, while the user might think it's one of the others. This leads to Pollack inconsistency.

Here is an instance reported recently by @markulf (where the first type for reverse is incorrect):
```
val reverse: #a:Type -> f:(list a -> Tot (list a)){forall l. l == f (f l)}
val reverse: list 'a -> Tot (list 'a)
let rec reverse l =
  match l with
  | [] -> []
  | hd::tl -> append (reverse tl) [hd]
```
Or for more effect, here is a "proof of false":
```
val inconsistent : unit -> Lemma False
val inconsistent : unit -> Lemma True
let inconsistent () = ()
```
Also noticed by @s-zanella in the past:
#829,
https://github.com/FStarLang/FStar/issues/965#issuecomment-293217030

**Comment by markulf** [comment]
When discussing this pollack inconsistency was mentioned a few times. Here is a link to the nice paper introducing it: http://www.cs.ru.nl/~freek/pubs/rap.pdf

**Comment by kyoDralliam** [comment]
Should be fixed by merging fb0fc3b4520b044052c5d9df2bb0c00f2c4b931b
